### ISSUE TITLE
Unused use statement rule

### ISSUE THREAD

**Issue by alexwhitman** [squizlabs/PHP_CodeSniffer#801]
Is there a rule to detect unused use statements?  For example:
```php
<?php

use Foo\Bar;
use Foo\Baz; // This should be detected as unused

class MyClass
{
    public function doSomething(Bar $bar)
    {
        // .....
    }
}
```
PHP-CS-Fixer has an `unused_use` fixer but I'd prefer to use the one tool if possible.

**Comment by aik099** [comment]
Nope.

For a questions you can use our Gitter channel: https://gitter.im/squizlabs/PHP_CodeSniffer

**Comment by klausi** [comment]
There is a sniff for that in Drupal's coder: https://www.drupal.org/project/coder

**Comment by ExplodingCabbage** [comment]
I just wanted this feature and was considering writing my own sniff. @klausi, as the owner of coder would you have any objection to PHPCS nicking the [UnusedUseStatementSniff](https://github.com/klausi/coder/blob/8.x-2.x/coder_sniffer/Drupal/Sniffs/Classes/UnusedUseStatementSniff.php) and putting it into the `generic` section of PHPCS? It looks pretty solid to my (non-expert) eye, and I'm sure that more people would be able to benefit from it if it were put into the core of PHPCS; I'd prefer not to replicate your work by writing my own to PR into here.

(If you're okay with it, I'm also interested to look over coder for other non-Drupal-specific sniffs that could happily live in PHPCS and start migrating them. It looks like you've got a lot of cool stuff over there!)

**Comment by klausi** [comment]
Coder is GPLv2+ licensed and PHPCS is MIT, so you need to get copyright permission from all authors of a sniff when you want to pull in sniffs from Coder to PHPCS for MIT relicensing. Which is totally fine with me, so you have my permission to do that. In the case of UnusedUseStatementSniff we also need to ask @alexpott (the only other author).

**Comment by alexpott** [comment]
Fine by me.

**Comment by patrickallaert** [comment]
Looks like @alexpott agreed, can it be included?
Thanks in advance.

**Comment by ExplodingCabbage** [comment]
@patrickallaert there's [already an in-progress PR to include this](https://github.com/squizlabs/PHP_CodeSniffer/issues/1106).

**Comment by TomasVotruba** [comment]
Alread in `no_unused_imports [@Symfony]`: https://github.com/friendsofphp/php-cs-fixer

**Comment by juuuuuu** [comment]
I think this issue can be closed.

**Comment by rodrigoprimo** [comment]
Just noting here that Slevomat also has a sniff called `SlevomatCodingStandard.Namespaces.UnusedUses` that checks for unused use statements: https://github.com/slevomat/coding-standard/blob/0cd4b30cc1037eca54091c188d260d570e61770c/doc/namespaces.md#slevomatcodingstandardnamespacesunuseduses-
